### PR TITLE
Fix Nano Nemotron text-only weight loading

### PR DESCRIPTION
**Pull request by Baekpica**
## Purpose

Fix Nano Nemotron VL weight loading when all multimodal prompt limits are zero, including `--language-model-only`. In this mode the multimodal tower modules are replaced by missing-stage placeholders, so adapter, vision, and audio checkpoint weights should be skipped instead of inspected or loaded.

AI-assisted contribution: this PR includes AI-assisted code changes reviewed by the human submitter.

## Test Plan

- Add a unit test that constructs a text-only Nano Nemotron VL shell and verifies multimodal weights are ignored while language model weights are forwarded.

## Test Result

- `../.venv/bin/python -m pytest tests/models/multimodal/test_nano_nemotron_vl.py -q`
- Result: `1 passed, 16 warnings in 0.61s`

### CODE REVIEW

**Comment by Baekpica** [comment]
The sound-encoder crash case is handled in the current PR head. Sound weights are skipped when multimodal loading is enabled but `self.sound_encoder` is `None`:

```python
if not load_multimodal_weights or self.sound_encoder is None:
    continue
```

This is covered by `test_nano_nemotron_vl_skips_sound_weights_without_sound_encoder`, which exercises an image-enabled multimodal config with no audio encoder while checkpoint sound weights are present.

Local validation:

```bash
../.venv/bin/python -m py_compile vllm/model_executor/models/nano_nemotron_vl.py tests/models/multimodal/test_nano_nemotron_vl.py
../.venv/bin/python -m pytest tests/models/multimodal/test_nano_nemotron_vl.py -q
```

Result: `2 passed` for the Nano Nemotron VL regression test file.

The remaining `pre-run-check` failure is the repository label/author gate (`ready`/`verified` label or 4+ merged PRs), not a code/test failure.

File: vllm/model_executor/models/nano_nemotron_vl.py [hunk 1]
```diff
@@ -1523,23 +1530,29 @@ def is_sound_weights(name: str) -> bool:
...
             elif is_vision_weights(name):
+                if not load_multimodal_weights:
+                    continue
                 # Convert: vision_model.radio_model.* → radio_model.*
                 hf_key = name[len("vision_model.") :]  # Remove "vision_model." prefix
                 vision_weights.append((hf_key, w))
             elif is_sound_weights(name):
-                assert self.sound_encoder is not None
+                if not load_multimodal_weights or self.sound_encoder is None:
+                    continue
```
Comment (netanel-haber) [inline vllm/model_executor/models/nano_nemotron_vl.py]:
```suggestion
                assert self.sound_encoder is not None
                if not load_multimodal_weights:
                    continue
```
I prefer keeping the assertions stricts

File: vllm/model_executor/models/nano_nemotron_vl.py [hunk 2]
```diff
@@ -1499,7 +1499,14 @@ def compute_logits(
...
 
     def load_weights(self, weights: Iterable[tuple[str, torch.Tensor]]):
-        adapter_dict = dict(self.mlp1.named_parameters())
+        mm_config = self.model_config.multimodal_config
+        load_multimodal_weights = not all(
+            mm_config.get_limit_per_prompt(modality) == 0
+            for modality in ("image", "video", "audio")
+        )
+        adapter_dict = (
+            dict(self.mlp1.named_parameters()) if load_multimodal_weights else {}
```
Comment (netanel-haber) [inline vllm/model_executor/models/nano_nemotron_vl.py]:
```suggestion
            dict(self.mlp1.named_parameters())
```
I think we can still have this populated even if `load_multimodal_weights` is `False`, no? That's what I would prefer, but I think this is fine if you think this is the better choice.

Comment (Baekpica) [reply vllm/model_executor/models/nano_nemotron_vl.py]: Applied in Baekpica/vllm@5cbdffa46. Agreed that `mlp1` exists independently of whether multimodal weights are loaded, so `adapter_dict` can be populated unconditionally.

I also kept multimodal weights skipped when all modalities are disabled, while preserving the strict `sound_encoder` assertion once multimodal loading is active.

Tested with:

```bash
python -m pytest tests/models/multimodal/test_nano_nemotron_vl.py -q
```

**Comment by netanel-haber** [comment]
@Baekpica - can I ask that you confirm you've run the ordinary model and flow with your current changes, and that you send a video+audio request with e.g. https://github.com/chthomos/video-media-samples/blob/master/big-buck-bunny-1080p-30sec.mp4 (and use_audio_in_video: true), and that you get sane output? 
This will be sanity test to see that nothing is broken.

**Comment by Baekpica** [comment]
@netanel-haber I ran the requested sanity checks against the current PR head and confirmed both the ordinary text-only flow and the video+audio `use_audio_in_video: true` flow work.

## Environment

- Machine: PGX-ThinkStation / NVIDIA GB10, 128 GB unified memory
- OS/runtime: Linux `6.17.0-1014-nvidia-aarch64`, `aarch64`
- NVIDIA driver: `595.58.03`
- Python: `3.12.12`
- PyTorch: `2.11.0+cu130`, `torch.version.cuda == 13.0`
- vLLM source/commit: PR head `5cbdffa467df30c4d4c3172db2b1add4f415ced4`
- vLLM installed version: `0.20.1rc1.dev96+g5cbdffa46.precompiled`
- Model path used from the repo/workspace root: `models/models/Nemotron-3-Nano-Omni-30B-A3B-Reasoning-NVFP4`
- Audio/video deps: installed `vllm[audio]`; aligned `flashinfer-python==0.6.9`, `flashinfer-cubin==0.6.9`, `flashinfer-jit-cache==0.6.9+cu130`

## Unit regression test

```bash
cd vllm
../.venv/bin/python -m pytest tests/models/multimodal/test_nano_nemotron_vl.py -q
```

Result:

```text
3 passed, 16 warnings in 2.47s
```

## Ordinary text-only serving flow

From the workspace root, I served the local model with `--language-model-only` using this command:

```bash
vllm serve models/models/Nemotron-3-Nano-Omni-30B-A3B-Reasoning-NVFP4 \
  --served-model-name nemotron_3_nano_omni \
  --host 0.0.0.0 \
  --port 8001 \
  --trust-remote-code \
  --reasoning-parser nemotron_v3 \
  --enable-auto-tool-choice \
  --tool-call-parser qwen3_coder \
  --kv-cache-dtype fp8 \
  --language-model-only \
  --max-model-len 262144 \
  --max-num-seqs 2 \
  --gpu-memory-utilization 0.92 \
  --moe-backend marlin
```

Observed startup logs included:

```text
All limits of multimodal modalities supported by the model are set to 0, running in text-only mode.
Starting vLLM server on http://0.0.0.0:8001
Application startup complete.
```

Validation:

```bash
curl http://127.0.0.1:8001/health
curl http://127.0.0.1:8001/v1/models
```

Both returned HTTP 200. I then sent a text-only chat completion request and got sane output:

```text
text-only sanity check passed
```

## Video + audio serving flow with `use_audio_in_video: true`

Downloaded the requested sample video from the raw GitHub URL:

```bash
curl -L --fail -o /tmp/big-buck-bunny-1080p-30sec.mp4 \
  https://raw.githubusercontent.com/chthomos/video-media-samples/master/big-buck-bunny-1080p-30sec.mp4
```

Verified the media has both video and audio streams:

```text
[('video', 'h264', 30.0), ('audio', 'aac', 30.0)]
```

From the workspace root, I served the same local model in multimodal mode using this command:

```bash
vllm serve models/models/Nemotron-3-Nano-Omni-30B-A3B-Reasoning-NVFP4 \
  --served-model-name nemotron_3_nano_omni \
  --host 0.0.0.0 \
  --port 8001 \
  --trust-remote-code \
  --reasoning-parser nemotron_v3 \
  --enable-auto-tool-choice \
  --tool-call-parser qwen3_coder \
  --kv-cache-dtype fp8 \
  --max-model-len 131072 \
  --max-num-seqs 8 \
  --gpu-memory-utilization 0.80 \
  --limit-mm-per-prompt '{"video":1,"image":1,"audio":1}' \
  --media-io-kwargs '{"video":{"fps":2,"num_frames":256}}' \
  --allowed-local-media-path / \
  --enable-prefix-caching \
  --max-num-batched-tokens 32768
```

Observed startup logs included:

```text
Multi-modal warmup completed
Readonly multi-modal warmup completed
Starting vLLM server on http://0.0.0.0:8001
Application startup complete.
```

Then I sent this request:

```json
{
  "model": "nemotron_3_nano_omni",
  "messages": [
    {
      "role": "user",
      "content": [
        {
          "type": "video_url",
          "video_url": {
            "url": "file:///tmp/big-buck-bunny-1080p-30sec.mp4"
          }
        },
        {
          "type": "text",
          "text": "Describe the video and mention what you hear. Keep the answer under two sentences."
        }
      ]
    }
  ],
  "max_tokens": 128,
  "temperature": 0.2,
  "top_k": 1,
  "chat_template_kwargs": {
    "enable_thinking": false
  },
  "mm_processor_kwargs": {
    "use_audio_in_video": true
  }
}
```

The server log confirmed the audio-in-video path was used:

```text
Using 1 pre-populated audio item(s) from upstream.
POST /v1/chat/completions HTTP/1.1 200 OK
```

The response was sane and referenced both the video and audio:

```text
The video opens with a serene animated landscape featuring a bird perched on a branch, followed by a title card for "The Peach Open Movie Project Presents" and "Big Buck Bunny," set against a backdrop of lush greenery and a tranquil stream. The scene is accompanied by gentle, melodic music and subtle nature sounds, creating a peaceful atmosphere.
```

**Comment by netanel-haber** [comment]
Excellent. I gave my LGTM. Let's wait for @tomeras91 [who has review/write permissions] to give his review now.

**Review by netanel-haber — APPROVED** [review]

**Review by tomeras91 — APPROVED** [review]
Thanks for the fix! LGTM

**Comment by Baekpica** [comment]
This looks unrelated to this PR. The failed job is `fusion-e2e-tp2-b200` with exit status 125 on the B200 queue, while this PR only touches Nano
Nemotron VL weight loading/tests. 

@tomeras91  Could you please rerun the failed Buildkite job? 🙏
